timedctl/timedctl.py

```diff
@@ -25,652 +24,527 @@
+        ]
+        # create a list for fzf
+        fzf_obj = []
+        for row in activity_view:
+            fzf_obj.append(
+                [" | ".join([row[0], row[1], row[2]]), row[1], row[2], row[3], row[4]],
+            )
+
+        activity_obj = fzf_wrapper(fzf_obj, [0], "Select an activity: ")
+        return activity_obj
+
+    def format_activity(self, activity_obj):
+        """Format an activity for display."""
+        task_obj = activity_obj["relationships"]["task"]
 
-    report = fzf_wrapper(fzf_obj, [0], "Select a report: ")
-    return report
+        task = task_obj["attributes"]["name"]
 
+        project_obj = self.timed.projects.get(
+            id=task_obj["relationships"]["project"]["id"],
+            cached=True,
+        )
+        project = project_obj["attributes"]["name"]
 
-def select_activity(date):
-    """FZF prompt to select an activity."""
-    activities = timed.activities.get(filters={"date": date})
-    activity_view = []
-    # loop through all activities
-    for activity_obj in activities:
-        # check if there is an actual task, else use an unknown task
-        task_data = activity_obj["relationships"]["task"]["data"]
-        if task_data:
-            task = timed.tasks.get(id=task_data["id"], cached=True)
+        customer_obj = self.timed.customers.get(
+            id=project_obj["relationships"]["customer"]["data"]["id"],
+            cached=True,
+        )
+        customer = customer_obj["attributes"]["name"]
+        return f"{customer} > {project} > {task}"
+
+    def get_customer_by_name(self, customers, name, archived):
+        """Get customer by name."""
+        customers = self.timed.customers.get(
+            cached=True,
+            filters={"archived": archived},
+        )
+        customer = [c for c in customers if c["attributes"]["name"] == name]
+        if len(customer) == 0:
+            error_handler("ERR_CUSTOMER_NOT_FOUND")
+        customer_id = customer[0]["id"]
+        return customer_id
+
+    def get_project_by_name(self, projects, name, customer_id, archived):
+        """Get project by name."""
+        projects = self.timed.projects.get(
+            cached=True,
+            filters={"customer": customer_id, "archived": archived},
+        )
+        project = [c for c in projects if c["attributes"]["name"] == name]
+        if len(project) == 0:
+            error_handler("ERR_PROJECT_NOT_FOUND")
+        project_id = project[0]["id"]
+        return project_id
+
+    def get_task_by_name(self, tasks, name, project_id, archived):
+        """Get task by name."""
+        tasks = self.timed.tasks.get(
+            cached=True,
+            filters={"project": project_id, "archived": archived},
+        )
+        task = [c for c in tasks if c["attributes"]["name"] == name]
+        if len(task) == 0:
+            error_handler("ERR_TASK_NOT_FOUND")
+        task_id = task[0]["id"]
+        return task_id
```

Comment:
This could be refactored to this:
```suggestion
      if (task := next((t for t in tasks if t["attributes"]["name"] == name), None)):
          return task["id"]
      error_handler("ERR_TASK_NOT_FOUND")
```

Since this is a common pattern (getting something by name) I would recommend extracting this logic into a readable function and using that

Reply:
resolved as of https://github.com/adfinis/timedctl/pull/45/commits/13850fa95bec5246868b900b3db68602ca6edbf0